app/code/Magento/CatalogInventory/Model/StockStateProvider.php

```diff
@@ -106,6 +106,10 @@ public function checkQuoteItemQty(StockItemInterface $stockItem, $qty, $summaryQ
     {
         $result = $this->objectFactory->create();
         $result->setHasError(false);
+        
+        $product = $this->productFactory->create();
+        $product->load($stockItem->getProductId());
+        $stockItem->setProduct($product);
```

Comment:
@orlangur 
This is weird, because the interface already loaded Magento\CatalogInventory\Model\Stock\Item
And there is a public function setProduct available.
In my system it is working fine, no errors.